### ISSUE TITLE
[Feature Request] Support adding response flags and response code details in custom response headers

### ISSUE THREAD

**Issue by AngelaNing97** [envoyproxy/envoy#11648]
Description:
Envoy supports adding a set of custom request/response headers and also dynamic values in those headers. By adding headers with RESPONSE_CODE_DETAILS and RESPONSE_FLAGS, we provide additional context to clients whether the response code has come from Envoy or the upstream backend, as well as why the request has failed.  

Envoy currently has some support for RESPONSE_CODE_DETAILS and RESPONSE_FLAGS as custom headers, but we observed that in cases where the Envoy sidecar sends a local 503 response, for example, when there are no healthy upstreams, custom headers are not getting added to the responses. Investigating the issue, we noticed that `finalizeResponseHeaders` are not called when the Envoy Router `sendLocalReply`. 

`callbacks_->sendLocalReply(Http::Code::ServiceUnavailable, "no healthy upstream", modify_headers_, absl::nullopt, StreamInfo::ResponseCodeDetails::get().NoHealthyUpstream);`

A possible solution might be to call `finalizeResponseHeaders` in `modify_headers_` for cases that Envoy sends a local reply.

[optional Relevant Links:]
Custom request/response header Envoy doc: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers
https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#config-access-log-format-response-flags

**Comment by htuch** [comment]
Seems plausible to me. I wonder if we could build control of this into the local reply mapper. @mattklein123 @alyssawilk

**Comment by mattklein123** [comment]
+1 we should add this. Also +1 for possibly controlling this behavior through the mapper. cc @qiwzhang

**Comment by alyssawilk** [comment]
SGTM - we've found it quite useful for debug as well.

Two suggestions.  First, it may be worth config for sending an enum:  "never send, send iff internal address, always send".  
Second, and I'd like @mattklein123 's thought on this one, I think especially for this, which is very Envoy-ish info, it may be worth sticking information in a single tokenized header.  Internally we have a bunch of stuff which can be configured to send for hop by hop debug info, and the over head of
debug-header-A: A value
debug-header-B: B value
when we have 15+ of these, can end up being an rtt with a small initial CWND.    I think it's worth glomming them together with debug-header: A;a-val,B-b-val,C;c-val with easy utils for composing and tokenizing individual values.

**Comment by mattklein123** [comment]
Yes agreed on supporting a tokenized header, though I suspect with our header formatters and the local reply config we could probably make this all pretty easily configurable?

**Comment by AngelaNing97** [comment]
The enum "never send, send iff internal address, always send" should probably be easily configurable in the `header_formatter`. My interpretation for the second part of @alyssawilk 's comment is that at each hop, we append the response_code_details, response_flags, or other desired debugging info to the "tokenized header" object? Is this correct? Could you please provide an example of what the tokenized header would look like in a real use case? Also, I am wondering if your team is interested in pick this up.

**Comment by alyssawilk** [comment]
yeah, as Matt says some of this can go in the header formatter.
I think that appending various debug info per hop would be fantastic, but I don't have cycles to pick this up yet so I think it'll stay help wanted until someone volunteers :-)

**Comment by rulex123** [comment]
Is anybody working on this? Otherwise I would like to give it a go :-)

**Comment by mattklein123** [comment]
@rulex123 go for it! I will assign you.

**Comment by mike1808** [comment]
This issue is also helpful for a chain of Envoys and when you want to implement retries on outer Envoy. For example, in Istio when you have one Envoy as an ingress gateway and other one as a sidecar and on your ingress gateway you want to retry only on HTTP errors and not on TCP errors you won't be able to, as the sidecar Envoy will just reply 503 HTTP error to the ingress gateway without any details why 503 happened. I've tried to add `%RESPONSE_FLAGS%` in the response headers of the sidecar, so if upstream fails the ingress gateway knew the reason and retries using `retriable_headers` condition.

**Comment by mike1808** [comment]
Also, I'm not sure if it's a [Feature Request] as there is no mentioning that `response_headers_to_add` do not work on 503 errors.

**Comment by mattklein123** [comment]
> Also, I'm not sure if it's a [Feature Request] as there is no mentioning that response_headers_to_add do not work on 503 errors.

This is covered by https://github.com/envoyproxy/envoy/issues/4317. cc @alyssawilk
